<launch>
	
	<arg name="sensor_type" default="range_bearing" />
	
  <arg name="rviz_on" default="true" />

<!--
	Robot Information
-->
  <arg name="agent_prefix" default="robot" />
  <arg name="num_agents"   default="1" />

  <arg name="sensor_config_file" default="$(find phd_sensor)/config/$(arg sensor_type)_sensor_dynamic.yaml" />

	<arg name="sensor_frame_id" default="sensor" />
	<arg name="base_frame_id"   default="base" />
  <arg name="odom_frame_id"   default="odom" />
  <arg name="global_frame_id" default="/map" />
  
<!--
  Target information
-->
  <arg name="num_target_types" default="4" />

  <arg name="target_prefix0" default="chair" />
	<arg name="num_targets0"   default="10" />
  <arg name="motion_model_file0"  value="$(find phd_filter)/config/static_motion_model.yaml" />
	
	<arg name="target_prefix1" default="table" />
	<arg name="num_targets1"   default="0" />
  <arg name="motion_model_file1"  value="$(find phd_filter)/config/static_motion_model.yaml" />
	
	<arg name="target_prefix2" default="person" />
	<arg name="num_targets2"   default="10" />
  <arg name="motion_model_file2"  value="$(find phd_filter)/config/grw_motion_model.yaml" />

	<arg name="target_prefix3" default="dog" />
	<arg name="num_targets3"   default="0" />
  <arg name="motion_model_file3"  value="$(find phd_filter)/config/grw_motion_model.yaml" />

<!--
	Sensor simulation
-->
	<include file="$(find phd_sensor)/launch/dynamic_sensor.launch" pass_all_args="true" >
		<arg name="rviz_on" value="false" />
	</include>
	
<!--
  RVIZ
-->
  <group if="$(arg rviz_on)">
    <node name="rviz" pkg="rviz" type="rviz"
          args="--display-config=$(find phd_filter)/launch/robot0_all.rviz" />
  </group>

<!--
	Aggregate robot sensor measurements for PHD filter
-->
  <node name="sensor_aggregator" pkg="phd_sensor" type="$(arg sensor_type)_aggregator"
        output="screen" respawn="false" launch-prefix="">
    <param name="num_agents" value="$(arg num_agents)"/>
    <param name="agent_prefix" value="$(arg agent_prefix)"/>
    <param name="topic_name" value="measurements_throttle"/>
  </node>

  
<!--
       Use only when calculating OSPA error within sensor Fov.
       Need also to change topic from "target_array" to "target_fov" 
       in both phd_filter.hpp and next node in this launch file.

  <node name="inside_fov" pkg="phd_filter" type="inside_fov" output="screen">
      		<remap from="gt_pose" to="/$(arg agent_prefix)0/gt_pose"/>
		<remap from="target_fov" to="/target_fov" />
		<param name="fov" value="10.0" />
  </node>
-->
<!--
	PHD filter
-->
  <node name="phd_filter" pkg="phd_filter" type="phd_filter_$(arg sensor_type)_node"
        output="screen" respawn="false" launch-prefix="">
		<remap from="map" to="/map" />
		<remap from="pose" to="gt_pose" />
		<remap from="measurements" to="/measurements" />
		<!--remap from="target_fov" to="/target_fov" /-->
		<remap from="target_array" to="/target_array" />
		<param name="num_targets" value="10.0" />
		<param name="max_value" value="0.05" />
		<param name="width" value="20.0" />
		<param name="height" value="20.0" />
		<param name="origin_x" value="1.0"/>
		<param name="origin_y" value="1.0"/>
		<param name="grid_res" value="0.20" />
		<param name="ospa_filename" value="/home/junchen/catkin_ws/src/phd-filter-range-bearing/phd_filter/OSPA_dynamic.txt" />

		<rosparam file="$(arg sensor_config_file)" ns="sensor" command="load" />
		<rosparam file="$(arg motion_model_file0)" ns="motion/$(arg target_prefix0)" command="load" />
		<rosparam file="$(arg motion_model_file1)" ns="motion/$(arg target_prefix1)" command="load" />
		<rosparam file="$(arg motion_model_file2)" ns="motion/$(arg target_prefix2)" command="load" />
  </node>

<include file="$(find phd_filter)/launch/robot_moving.launch" >
	<arg name="moving_type" value="$(arg agent_prefix)0" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)0" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)1" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)2" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)3" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)4" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)5" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)6" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)7" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)8" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)9" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<!--include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)0" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)1" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)2" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)3" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)4" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)5" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)6" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)7" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)8" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/dog_moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix3)9" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include-->

<!--include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)10" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)11" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)12" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)13" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)14" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)15" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)16" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)17" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)18" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)19" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)20" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)21" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)22" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)23" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)24" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)25" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)26" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)27" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)28" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include>

<include file="$(find phd_filter)/launch/moving.launch" >
	<arg name="moving_type" value="$(arg target_prefix2)29" />
        <arg name="base_frame_id"   default="base" />
        <arg name="global_frame_id" default="/map" />
</include-->

</launch>
